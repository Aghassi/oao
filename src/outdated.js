// @flow

import { readAllSpecs } from './utils/readSpecs';
import removeInternalLinks from './utils/removeInternalLinks';
import writeSpecs from './utils/writeSpecs';
import { exec } from './utils/shell';

type Options = {
  src: string,
  link: ?string,
};

const run = async (opts: Options) => {
  const { src: srcPatterns, link: linkPattern } = opts;
  const allSpecs = await readAllSpecs(srcPatterns);
  const pkgNames = Object.keys(allSpecs);
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    const { pkgPath, specPath, specs: prevSpecs } = allSpecs[pkgName];

    // Rewrite package.json without own/linked packages, run `yarn outdated`, and revert changes
    let fModified = false;
    try {
      const { nextSpecs } = removeInternalLinks(prevSpecs, pkgNames, linkPattern);
      if (nextSpecs !== prevSpecs) {
        writeSpecs(specPath, nextSpecs);
        fModified = true;
      }
      await exec('yarn outdated', { cwd: pkgPath });
    } finally {
      if (prevSpecs != null && fModified) writeSpecs(specPath, prevSpecs);
    }
  }
};

export default run;
