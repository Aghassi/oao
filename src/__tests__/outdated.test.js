/* eslint-env jest */
/* eslint-disable global-require, import/no-dynamic-require */

import path from 'path';
import outdated from '../outdated';

jest.mock('../utils/shell');
jest.mock('../utils/writeSpecs');

const PACKAGE_NAMES_1 = ['oao', 'oao-b', 'oao-c', 'oao-d', 'oao-priv'];

const readOriginalSpecs = (base, names) => {
  const originalSpecs = {};
  names.forEach((name) => {
    originalSpecs[name] = require(path.join(base, `${name}/package.json`));
  });
  return originalSpecs;
};

const spyFinalSpec = (spy, pkgName) => {
  let finalSpec;
  spy.mock.calls.forEach(([, specs]) => {
    if (specs.name === pkgName) finalSpec = specs;
  });
  return finalSpec;
};

describe('OUTDATED command', () => {
  beforeEach(() => { jest.resetAllMocks(); });

  it('does not modify any package.json', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    const base = path.join(process.cwd(), 'test/fixtures/packages');
    const originalSpecs = readOriginalSpecs(base, PACKAGE_NAMES_1);
    await outdated({ src: 'test/fixtures/packages/*' });
    // Process call arguments; we keep the last time a spec is stored
    // and compare it with the original one
    writeSpecs.mock.calls.forEach(([specPath, specs]) => {
      const { name } = specs;
      expect(specPath.split(path.sep)).toContain(name);
      expect(specPath.split(path.sep)).toContain('package.json');
    });
    PACKAGE_NAMES_1.forEach((name) => {
      const finalSpecWritten = spyFinalSpec(writeSpecs, name);
      if (finalSpecWritten === undefined) return;  // not changed at all!
      expect(finalSpecWritten).toEqual(originalSpecs[name]);
    });
  });

  it('executes the correct `yarn outdated`s', async () => {
    const helpers = require('../utils/shell');
    await outdated({ src: 'test/fixtures/packages/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });
});
