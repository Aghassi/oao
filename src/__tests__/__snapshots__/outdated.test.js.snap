// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OUTDATED command executes the correct \`yarn outdated\`s 1`] = `
Array [
  Array [
    "yarn outdated",
    Object {
      "cwd": "test/fixtures/packages/oao",
    },
  ],
  Array [
    "yarn outdated",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
    },
  ],
  Array [
    "yarn outdated",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
    },
  ],
  Array [
    "yarn outdated",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
    },
  ],
  Array [
    "yarn outdated",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
    },
  ],
  Array [
    "yarn outdated",
    Object {
      "cwd": ".",
    },
  ],
]
`;
